test/jdk/sun/security/pkcs11/fips/NssdbPin.java

```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        /*
+         * Functional cross-test against an NSS DB generated by modutil
+         * with the same PIN. Check that we can perform a crypto operation
+         * that requires a login. The login might be explicit or implicit.
+         */
+        Provider p = Security.getProvider(FIPS_PROVIDER_NAME);
+        if (DEBUG) {
+            System.out.println(FIPS_PROVIDER_NAME + ": " + p);
+        }
+        if (p == null) {
+            throw new Exception(FIPS_PROVIDER_NAME + " initialization failed.");
+        }
+        if (DEBUG) {
+            System.out.println("Login type: " + loginType);
+        }
+        if (loginType == LoginType.EXPLICIT) {
+            // Do the expansion to account for truncation, so C_Login in
+            // the NSS Software Token gets a UTF-8 encoded PIN.
+            byte[] pinUtf8 = expectedPin.getBytes(StandardCharsets.UTF_8);
+            char[] pinChar = new char[pinUtf8.length];
+            for (int i = 0; i < pinChar.length; i++) {
+                pinChar[i] = (char)(pinUtf8[i] & 0xFF);
+            }
+            KeyStore.getInstance("PKCS11", p).load(null, pinChar);
+            if (DEBUG) {
+                System.out.println("Explicit login succeeded.");
+            }
+        }
+        if (DEBUG) {
+            System.out.println("Trying a crypto operation...");
+        }
+        final int blockSize = 16;
+        Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding", p);
+        cipher.init(Cipher.ENCRYPT_MODE,
+                new SecretKeySpec(new byte[blockSize], "AES"));
+        if (cipher.doFinal(new byte[blockSize]).length != blockSize) {
+            throw new Exception("Could not perform a crypto operation.");
+        }
+        if (DEBUG) {
+            if (loginType == LoginType.IMPLICIT) {
+                System.out.println("Implicit login succeeded.");
+            }
+            System.out.println("Crypto operation after login succeeded.");
+        }
+
+        if (loginType == LoginType.IMPLICIT) {
+            /*
+             * Additional unitary testing. Expected to succeed at this point.
+             */
+            if (DEBUG) {
+                System.out.println("Trying unitary test...");
+            }
+            String sysPathProp = System.getProperty(FIPS_NSSDB_PATH_PROP);
+            if (DEBUG) {
+                System.out.println("Path value (as a System property): " +
+                        sysPathProp);
+            }
+            if (!expectedPath.equals(sysPathProp)) {
+                throw new Exception("Path is different than expected: " +
+                        sysPathProp + " (actual) vs " + expectedPath +
+                        " (expected).");
+            }
+            Class<?> c = Class
+                    .forName("sun.security.pkcs11.FIPSTokenLoginHandler");
+            Method m = c.getDeclaredMethod("getFipsNssdbPin");
+            m.setAccessible(true);
+            String pin = null;
```

Comment:
Maybe I'm missing something, but couldn't we just have declared `String pin = "";` instead of `String pin = null;` here?

Wouldn't it have allowed us to keep the `!pin.equals(expectedPin)` previous (simpler) condition in the following `if` block? NOTE: with 'previous' I mean the left side of f9afab177d4b09e1e23a79ebe311d978101c416a...69cabe49b1802876a4b277ea2f99742245436652.

Reply:
The `pin` local variable initialized to `null` allows us to distinguish whether the `getFipsNssdbPin` returned empty string or null, which is the change that we just made and the behavior that we expect.